SgmlReaderDll/SgmlReader.cs

```diff
@@ -1951,17 +2019,68 @@ private bool ParseText(char ch, bool newtext)
                 else
                 {
                     if (!_current.IsWhitespace)
+                    {
                         ws = false;
+                    }
+                            
                     _sb.Append(ch);
                     ch = _current.ReadChar();
                 }
             }
 
-            string value = _sb.ToString();
+            string value;
+            if(_textWhitespace == TextWhitespaceHandling.None)
+            {
+                value = _sb.ToString();
+            }
+            else
+            {
+                //TextWhitespaceHandling handling = _textWhitespace;
+                //if (!newtext) handling &= ~TextWhitespaceHandling.TrimLeading; // Don't trim leading whitespace if this is not at the start of a run. UPDATE: Ah, I forget that `_sb` is persisted.
+                value = TrimStringBuilder(_sb, _textWhitespace);
+            }
+
             Push(null, XmlNodeType.Text, value);
             return ws;
         }
 
+        private static string TrimStringBuilder(StringBuilder sb, TextWhitespaceHandling handling) // It's simpler to return a substring from within a StringBuilder than it is to prevent whitespace from being added in the first place, hence this approach.
+        {
+            int startIndex = 0;
+            int endIndex   = sb.Length - 1; // `endIndex` is inclusive, not exclusive.
+
+            bool onlyLineBreaks = (handling & TextWhitespaceHandling.OnlyLineBreaks) == TextWhitespaceHandling.OnlyLineBreaks;
+           
+            if ((handling & TextWhitespaceHandling.TrimLeading) == TextWhitespaceHandling.TrimLeading)
```

Comment:
Perhaps cache these expressions as member booleans if this these if tests ends up generating a lot of IL code...

Reply:
They're bitwise operations though, so they should be very fast after the JIT does its thing - but I'll try your suggestion as that would _at least_ help readability.

I was originally going to use `.HasFlag()` but `.HasFlag()` is actually a lot slower than bitwise operations as it does a type-check every time, which is disappointing.

Reply:
Yep only if it makes a perf difference.  Memory usage of SgmlReader is not an issue, no one is every going to create millions of them at once.  So it's the right trade off to make if it helps.